modules/parquet/src/lib/parsers/parse-parquet-wasm.ts

```diff
@@ -3,32 +3,70 @@
 // Copyright (c) vis.gl contributors
 
 // eslint-disable
-import type {ArrowTable} from '@loaders.gl/arrow';
+import type {ReadableFile} from '@loaders.gl/loader-utils';
+import type {ArrowTable, ArrowTableBatch, Schema} from '@loaders.gl/schema';
 import {serializeArrowSchema} from '@loaders.gl/arrow';
-import type {ParquetWasmLoaderOptions} from '../../parquet-wasm-loader';
-import {loadWasm} from '../utils/load-wasm';
+
+import type * as parquetWasm from 'parquet-wasm';
 import * as arrow from 'apache-arrow';
 
-export async function parseParquetWasm(
-  arrayBuffer: ArrayBuffer,
-  options: ParquetWasmLoaderOptions
+import {loadWasm} from '../utils/load-wasm';
+import {makeStreamIterator} from '../utils/make-stream-iterator';
+
+export async function parseParquetFileWasm(
+  file: ReadableFile,
+  options?: parquetWasm.ReaderOptions & {wasmUrl: string}
 ): Promise<ArrowTable> {
-  const arr = new Uint8Array(arrayBuffer);
+  const wasmUrl = options?.wasmUrl;
+
+  const wasm = await loadWasm(wasmUrl);
+
+  let parquetFile: parquetWasm.ParquetFile;
+  if (file.handle instanceof Blob) {
+    // TODO - let's assume fromFile() works on Blobs and not just on File...
+    parquetFile = await wasm.ParquetFile.fromFile(file.handle as File);
+  } else {
+    parquetFile = await wasm.ParquetFile.fromUrl(file.url);
+  }
 
-  const wasmUrl = options?.parquet?.wasmUrl;
-  try {
-    const wasm = await loadWasm(wasmUrl);
-    const wasmTable = wasm.readParquet(arr);
+  const wasmTable = await parquetFile.read(options);
+  const ipcStream = wasmTable.intoIPCStream();
+  const arrowTable = arrow.tableFromIPC(ipcStream);
+
+  return {
+    shape: 'arrow-table',
+    schema: serializeArrowSchema(arrowTable.schema),
+    data: arrowTable
+  };
+}
+
+export async function* parseParquetFileInBatchesWasm(
+  file: ReadableFile,
+  options: parquetWasm.ReaderOptions & {wasmUrl: string}
+): AsyncIterable<ArrowTableBatch> {
+  const wasmUrl = options?.wasmUrl;
+
+  const wasm = await loadWasm(wasmUrl);
+
+  let parquetFile: parquetWasm.ParquetFile;
+  if (file.handle instanceof Blob) {
+    // TBD - does this work on Blob or just on File?
+    parquetFile = await wasm.ParquetFile.fromFile(file.handle as File);
+  } else {
+    parquetFile = await wasm.ParquetFile.fromUrl(file.url);
+  }
 
-    const ipcStream = wasmTable.intoIPCStream();
-    const arrowTable = arrow.tableFromIPC(ipcStream);
+  const stream: ReadableStream<arrow.Table> = await parquetFile.stream(options);
```

Comment:
This creates a stream of [wasm RecordBatch objects](https://kylebarron.dev/parquet-wasm/classes/esm_parquet_wasm.RecordBatch.html). You need to manage the movement to JS for each batch yourself. You can use the [`intoIPCStream` method](https://kylebarron.dev/parquet-wasm/classes/esm_parquet_wasm.RecordBatch.html#intoIPCStream) plus `arrow.tableFromIPC`

Reply:
> This creates a stream of [wasm RecordBatch objects](https://kylebarron.dev/parquet-wasm/classes/esm_parquet_wasm.RecordBatch.html). You need to manage the movement to JS for each batch yourself. 

Thanks, looks very promising indeed, will come back to this in separate PR as integrating it will need a bit more plumbing in loaders.gl

Nit: Any chance you could add a method that returns an async iterator so that I don't have to wrap things? Generally the Stream callback API is more fiddly to work with and AFAIK, stream incompatibilities between Node and browser are still a thing.

Reply:
I agree that looks annoying. 

Right now I'm using the [wasm-streams](https://github.com/MattiasBuelens/wasm-streams) library to return a `ReadableStream`. It doesn't look like there's any built-in support for returning an async iterator.

I have a note for myself (https://github.com/kylebarron/parquet-wasm/issues/307) to point users to a polyfill. In testing https://observablehq.com/d/f5723cea6661fb71 a long time ago, I used [this minimal polyfill](https://jakearchibald.com/2017/async-iterators-and-generators/#a-shorter-implementation), but I don't really understand what's happening

Reply:
Yes I suspected you were relying on some wasm support library. The polyfill you copies is similar to but cleaner than the one I added: https://github.com/visgl/loaders.gl/blob/master/modules/parquet/src/lib/utils/make-stream-iterator.ts.

> but I don't really understand what's happening

Yes that is the problem - converting between streams and iterators is way too obscure for most users. 
- At least Node streams define `[Symbol.asyncIterator]` and can be used as iterators, but the browser WhatWG streams can not.